fog/view/enclave/impl/src/oblivious_utils.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Contains methods that allow a Fog View Router enclave to combine all of the
+//! Fog View Shard's query responses into one query response that'll be returned
+//! for the client.
+
+use crate::Result;
+
+use aligned_cmov::{
+    subtle::{Choice, ConditionallySelectable, ConstantTimeEq},
+    CMov,
+};
+use alloc::{vec, vec::Vec};
+use mc_fog_types::view::{TxOutSearchResult, TxOutSearchResultCode};
+
+/// The default TxOutSearchResultCode used when collating the shard responses.
+///   Warning: Do not change this without careful thought because the logic in
+///            the [should_over_write_tx_out_search_result] method assumes that
+///            the default code is NotFound.
+const DEFAULT_TX_OUT_SEARCH_RESULT_CODE: TxOutSearchResultCode = TxOutSearchResultCode::NotFound;
+/// The length of the ciphertext. If this changes because we add more fields to
+/// the ETxOutRecord struct, then this needs to be changed.
+const CIPHERTEXT_LENGTH: usize = 237;
+
+#[allow(dead_code)]
+pub fn collate_shard_tx_out_search_results(
+    client_search_keys: Vec<Vec<u8>>,
+    shard_tx_out_search_results: Vec<TxOutSearchResult>,
+) -> Result<Vec<TxOutSearchResult>> {
+    let mut client_tx_out_search_results: Vec<TxOutSearchResult> = client_search_keys
+        .iter()
+        .map(|client_search_key| TxOutSearchResult {
+            search_key: client_search_key.to_vec(),
+            result_code: DEFAULT_TX_OUT_SEARCH_RESULT_CODE as u32,
+            ciphertext: vec![0u8; CIPHERTEXT_LENGTH],
+        })
+        .collect();
+
+    for shard_tx_out_search_result in shard_tx_out_search_results.iter() {
+        for client_tx_out_search_result in client_tx_out_search_results.iter_mut() {
+            maybe_overwrite_tx_out_search_result(
+                client_tx_out_search_result,
+                shard_tx_out_search_result,
+            );
+        }
+    }
+
+    Ok(client_tx_out_search_results)
+}
+
+fn maybe_overwrite_tx_out_search_result(
+    client_tx_out_search_result: &mut TxOutSearchResult,
+    shard_tx_out_search_result: &TxOutSearchResult,
+) {
+    let should_overwrite_tx_out_search_result = should_overwrite_tx_out_search_result(
+        client_tx_out_search_result,
+        shard_tx_out_search_result,
+    );
+    let shard_ciphertext_length = shard_tx_out_search_result.ciphertext.len();
+    assert_eq!(
+        shard_ciphertext_length, CIPHERTEXT_LENGTH,
+        "Shard ciphertext has unexpected length"
```

Comment:
i think this isn't going to be true in prod, because we've already changed the ciphertext length in prod several times:

* we added memos
* we added token ids

so there are currently shorter ciphertexts than this in the production database

So we have to figure out a way to factor out this assumption

Reply:
As discussed offline, I modified the client ciphertext as follows: the first byte is the `delta` between 255 and the shard ciphertext length, and the rest of the bits are filled with the shard ciphertext data until it runs out (and the rest is zero padded).